src/NCalc.Core/Parser/LogicalExpressionParser.cs

```diff
@@ -479,7 +407,78 @@ static LogicalExpressionParser()
         Parser = enableParserCompilation ? expressionParser.Compile() : expressionParser;
     }
 
-    private static LogicalExpression GetBinaryExpression((LogicalExpression, IReadOnlyList<(BinaryExpressionType, LogicalExpression)>) x)
+    private static LogicalExpression ParseNumber(ParseContext context, (long, int, long?, long?) number)
+    {
+        long integralValue = number.Item1;
+        int zeroCount = number.Item2;
+        long? decimalPart = number.Item3;
+        long? exponentPart = number.Item4;
+
+        if (((LogicalExpressionParserContext)context).Options.HasFlag(ExpressionOptions.DecimalAsDefault))
+        {
+            decimal result1 = integralValue;
+
+            // decimal part?
+            if (decimalPart != null && decimalPart.Value != 0)
+            {
+                var digits = Math.Floor(Math.Log10(decimalPart.Value) + 1) + zeroCount;
+                result1 += decimalPart.Value / (decimal)Math.Pow(10, digits);
+            }
+
+            // exponent part?
+            if (exponentPart == null) return new ValueExpression(result1);
+            var left = BigDecimal.Parse((double)result1);
+            var right = BigDecimal.Pow(10, exponentPart.Value);
+
+            var res = BigDecimal.Multiply(left, right);
+
+            if (res > decimal.MaxValue)
+                // There is no decimal.PositiveInfinity
+                return new ValueExpression(double.PositiveInfinity);
+
+            if (res < decimal.MinValue)
+                // There is no decimal.NegativeInfinity
+                return new ValueExpression(double.NegativeInfinity);
+
+            result1 = (decimal)res;
+
+            return new ValueExpression(result1);
+        }
+
+        double result = integralValue;
+
+        // decimal part?
+        if (decimalPart != null && decimalPart.Value != 0)
+        {
+            var digits = Math.Floor(Math.Log10(decimalPart.Value) + 1) + zeroCount;
+            result += decimalPart.Value / Math.Pow(10, digits);
+        }
+
+        // exponent part?
+        if (exponentPart != null)
+        {
+            var left = BigDecimal.Parse(result);
+            var right = BigDecimal.Pow(10, exponentPart.Value);
+
+            var res = BigDecimal.Multiply(left, right);
+
+            if (res > double.MaxValue)
+                result = double.PositiveInfinity;
+            else if (res < double.MinValue)
+                result = double.NegativeInfinity;
+            else
+                result = (double)res;
+        }
+
+        if (decimalPart != null || exponentPart != null)
+        {
+            return new ValueExpression(result);
+        }
+
+        return new ValueExpression((long)result);
```

Comment:
Missed this in previous review, can you please just return `integralValue` here?

Reply:
Good catch